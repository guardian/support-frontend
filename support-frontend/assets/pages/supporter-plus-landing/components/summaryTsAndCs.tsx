import { css } from '@emotion/react';
import { neutral, space, textSans17 } from '@guardian/source/foundations';
import { formatAmount } from 'helpers/forms/checkouts';
import type { IsoCurrency } from 'helpers/internationalisation/currency';
import {
	currencies,
	spokenCurrencies,
} from 'helpers/internationalisation/currency';
import type {
	ActiveProductKey,
	ActiveRatePlanKey,
} from 'helpers/productCatalog';
import { productCatalogDescription } from 'helpers/productCatalog';
import {
	BillingPeriod,
	getBillingPeriodNoun,
	ratePlanToBillingPeriod,
} from 'helpers/productPrice/billingPeriods';
import {
	getDateWithOrdinal,
	getLongMonth,
} from 'helpers/utilities/dateFormatting';
import { isSundayOnlyNewspaperSub } from 'pages/[countryGroupId]/helpers/isSundayOnlyNewspaperSub';

const containerSummaryTsCs = css`
	margin-top: ${space[6]}px;
	border-radius: ${space[2]}px;
	background-color: ${neutral[97]};
	padding: ${space[3]}px;
	${textSans17};
	color: ${neutral[0]};
	& a {
		color: ${neutral[7]};
	}
`;
export interface SummaryTsAndCsProps {
	productKey: ActiveProductKey;
	ratePlanKey: ActiveRatePlanKey;
	currency: IsoCurrency;
	amount: number;
}
export function SummaryTsAndCs({
	productKey,
	ratePlanKey,
	currency,
	amount,
}: SummaryTsAndCsProps): JSX.Element | null {
	const billingPeriod = ratePlanToBillingPeriod(ratePlanKey);
	const periodNoun = getBillingPeriodNoun(billingPeriod);
	const today = new Date();
	const renewalDateStart = `on the ${getDateWithOrdinal(today)} day of `;
	const renewalDateEnd = `every ${periodNoun}`;
	const renewalFrequency = `${renewalDateStart}${
		billingPeriod === BillingPeriod.Annual ? getLongMonth(today) + ' ' : ''
	}${renewalDateEnd}`;

	const isSundayOnlynewsletterSubscription = isSundayOnlyNewspaperSub(
		productKey,
		ratePlanKey,
	);

	if (isSundayOnlynewsletterSubscription) {
		return (
			<div css={containerSummaryTsCs}>
				The Observer subscription will auto renew each month. You will be
				charged the subscription amounts using your chosen payment method at
				each renewal, at the rate then in effect, unless you cancel.
			</div>
		);
	}

	const amountWithCurrency = formatAmount(
		currencies[currency],
		spokenCurrencies[currency],
		amount,
		false,
	);
	const summaryTsAndCsTierThreeGuardianAdLite = (
		<div css={containerSummaryTsCs}>
			The {productCatalogDescription[productKey].label} subscription
			{productKey === 'TierThree' ? 's' : ''} will auto-renew each {periodNoun}.
			You will be charged the subscription amount using your chosen payment
			method at each renewal, at the rate then in effect, unless you cancel.
		</div>
	);
	const summaryTsAndCs: Partial<Record<ActiveProductKey, JSX.Element>> = {
		Contribution: (
			<div css={containerSummaryTsCs}>
				We will attempt to take payment of {amountWithCurrency},{' '}
				{renewalFrequency}, from now until you cancel your payment. Payments may
				take up to 6 days to be recorded in your bank account. You can change
				how much you give or cancel your payment at any time.
			</div>
		),
		SupporterPlus: (
			<div css={containerSummaryTsCs}>
				The {productCatalogDescription[productKey].label} subscription and any
				contribution will auto-renew each {periodNoun}. You will be charged the
				subscription and contribution amounts using your chosen payment method
				at each renewal, at the rate then in effect, unless you cancel.
			</div>
		),
		TierThree: summaryTsAndCsTierThreeGuardianAdLite,
		GuardianAdLite: summaryTsAndCsTierThreeGuardianAdLite,
	};
	return summaryTsAndCs[productKey] ?? null;
}
