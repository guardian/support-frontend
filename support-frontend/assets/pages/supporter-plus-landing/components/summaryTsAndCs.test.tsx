import { render } from '@testing-library/react';
import type {
	ActiveProductKey,
	ActiveRatePlanKey,
} from 'helpers/productCatalog';
import { SummaryTsAndCs } from './summaryTsAndCs';

// Mocking price retrieval from productCatalog (not available in window at runtime)
jest.mock('helpers/utilities/dateFormatting', () => ({
	getDateWithOrdinal: () => 'first',
	getLongMonth: () => 'March',
}));

const ratePlanDescription: Partial<
	Record<ActiveRatePlanKey, string | undefined>
> = {
	WeekendPlus: 'Weekend Plus',
	SixdayPlus: 'Six Day Plus',
	Sunday: 'The Observer',
};

describe('Summary Ts&Cs Snapshot comparison', () => {
	it.each`
		productKey               | activeRatePlanKey
		${'Contribution'}        | ${'Monthly'}
		${'Contribution'}        | ${'Annual'}
		${'SupporterPlus'}       | ${'Monthly'}
		${'SupporterPlus'}       | ${'Annual'}
		${'TierThree'}           | ${'DomesticMonthly'}
		${'TierThree'}           | ${'DomesticAnnual'}
		${'OneTimeContribution'} | ${'OneTime'}
		${'GuardianAdLite'}      | ${'Monthly'}
		${'GuardianAdLite'}      | ${'Annual'}
		${'DigitalSubscription'} | ${'Monthly'}
		${'SubscriptionCard'}    | ${'WeekendPlus'}
		${'HomeDelivery'}        | ${'SixdayPlus'}
		${'SubscriptionCard'}    | ${'Sunday'}
		${'HomeDelivery'}        | ${'Sunday'}
	`(
		`summaryTs&Cs for $productKey With ratePlanKey $activeRatePlanKey renders correctly`,
		({ productKey, activeRatePlanKey }) => {
			const { container } = render(
				<SummaryTsAndCs
					productKey={productKey as ActiveProductKey}
					ratePlanKey={activeRatePlanKey as ActiveRatePlanKey}
					ratePlanDescription={
						ratePlanDescription[activeRatePlanKey as ActiveRatePlanKey]
					}
					currency={'GBP'}
					amount={0}
					isPaperProductTest={
						!!['WeekendPlus', 'SixdayPlus'].includes(
							activeRatePlanKey as ActiveRatePlanKey,
						)
					}
				/>,
			);
			expect(container.textContent).toMatchSnapshot();
		},
	);
});
