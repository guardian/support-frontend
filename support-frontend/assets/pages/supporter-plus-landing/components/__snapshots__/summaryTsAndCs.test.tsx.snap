// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary Ts&Cs Snapshot comparison renders summaryTs&Cs for the Digital SUbscription when the Premium Digital flag is enabled 1`] = `"The Premium digital subscription will auto-renew each month. You will be charged the subscription amount using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for Contribution With ratePlanKey Annual renders correctly 1`] = `"We will attempt to take payment of £0, on the first day of March every year, from now until you cancel your payment. Payments may take up to 6 days to be recorded in your bank account. You can change how much you give or cancel your payment at any time."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for Contribution With ratePlanKey Monthly renders correctly 1`] = `"We will attempt to take payment of £0, on the first day of every month, from now until you cancel your payment. Payments may take up to 6 days to be recorded in your bank account. You can change how much you give or cancel your payment at any time."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for DigitalSubscription With ratePlanKey Monthly renders correctly 1`] = `""`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for GuardianAdLite With ratePlanKey Annual renders correctly 1`] = `"The Guardian Ad-Lite subscription will auto-renew each year. You will be charged the subscription amount using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for GuardianAdLite With ratePlanKey Monthly renders correctly 1`] = `"The Guardian Ad-Lite subscription will auto-renew each month. You will be charged the subscription amount using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for HomeDelivery With ratePlanKey SixdayPlus renders correctly 1`] = `"The Six Day Plus subscription will auto renew each month. You will be charged the subscription amounts using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for HomeDelivery With ratePlanKey Sunday renders correctly 1`] = `"The Observer subscription will auto renew each month. You will be charged the subscription amounts using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for OneTimeContribution With ratePlanKey OneTime renders correctly 1`] = `""`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for SubscriptionCard With ratePlanKey Sunday renders correctly 1`] = `"The Observer subscription will auto renew each month. You will be charged the subscription amounts using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for SubscriptionCard With ratePlanKey WeekendPlus renders correctly 1`] = `"The Weekend Plus subscription will auto renew each month. You will be charged the subscription amounts using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for SupporterPlus With ratePlanKey Annual renders correctly 1`] = `"The All-access digital subscription and any contribution will auto-renew each year. You will be charged the subscription and contribution amounts using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for SupporterPlus With ratePlanKey Monthly renders correctly 1`] = `"The All-access digital subscription and any contribution will auto-renew each month. You will be charged the subscription and contribution amounts using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for TierThree With ratePlanKey DomesticAnnual renders correctly 1`] = `"The Digital + print subscriptions will auto-renew each year. You will be charged the subscription amount using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;

exports[`Summary Ts&Cs Snapshot comparison summaryTs&Cs for TierThree With ratePlanKey DomesticMonthly renders correctly 1`] = `"The Digital + print subscriptions will auto-renew each month. You will be charged the subscription amount using your chosen payment method at each renewal, at the rate then in effect, unless you cancel."`;
